File: packages/sdk/src/core/queue/offline-queue.ts
import { getItem, removeItem, setItem } from '../storage/storage';
import type { BatchItem } from '../types';
import { STORAGE_KEYS } from '../types';
import { logger } from '../utils/logger';

const MAX_QUEUE_SIZE = 5000;

export class OfflineQueue {
  private queue: BatchItem[] = [];
  private clientOrder = 0;
  private operationQueue = Promise.resolve();
  private initialized = false;

  async initialize(): Promise<void> {
    const result = await getItem<BatchItem[]>(STORAGE_KEYS.OFFLINE_QUEUE);
    if (
      result.success &&
      result.data &&
      Array.isArray(result.data) &&
      result.data.length > 0
    ) {
      this.queue = result.data;
      const orders = result.data
        .map((item) => item.clientOrder)
        .filter((order): order is number => typeof order === 'number');

      this.clientOrder = orders.length > 0 ? Math.max(...orders) + 1 : 0;
    }
    this.initialized = true;
  }

  enqueue(item: Omit<BatchItem, 'clientOrder'> | BatchItem): Promise<void> {
    if (!this.initialized) {
      logger.error('OfflineQueue not initialized, call initialize() first');
      return Promise.resolve();
    }

    const previousOp = this.operationQueue;
    this.operationQueue = previousOp.then(async () => {
      if (this.queue.length >= MAX_QUEUE_SIZE) {
        this.dropOldestSession();
      }

      const isRequeue =
        'clientOrder' in item && typeof item.clientOrder === 'number';

      const queueItem: BatchItem = isRequeue
        ? (item as BatchItem)
        : ({ ...item, clientOrder: this.clientOrder++ } as BatchItem);

      this.queue.push(queueItem);

      const result = await this.persist();
      if (!result.success) {
        logger.error(
          'Failed to persist queue after enqueue, data may be lost on crash'
        );
      }
    });
    return this.operationQueue;
  }

  private dropOldestSession(): void {
    const sessionItems = this.queue.filter((item) => item.type === 'session');

    if (sessionItems.length === 0) {
      const dropped = this.queue.shift();
      logger.debug(
        `Queue full (${MAX_QUEUE_SIZE}), no sessions to drop, dropping oldest item`,
        {
          type: dropped?.type,
          clientOrder: dropped?.clientOrder,
        }
      );
      return;
    }

    const oldestSession = sessionItems.reduce((oldest, current) =>
      current.clientOrder < oldest.clientOrder ? current : oldest
    );

    const oldestSessionId = oldestSession.payload.sessionId;

    const initialLength = this.queue.length;
    this.queue = this.queue.filter((item) => {
      if (item.type === 'device') {
        return true;
      }

      if (item.type === 'session') {
        return item.payload.sessionId !== oldestSessionId;
      }

      if (item.type === 'event' || item.type === 'ping') {
        return item.payload.sessionId !== oldestSessionId;
      }

      return true;
    });

    const droppedCount = initialLength - this.queue.length;
    logger.debug(
      `Queue full (${MAX_QUEUE_SIZE}), dropped oldest session and related items`,
      {
        sessionId: oldestSessionId,
        droppedCount,
      }
    );
  }

  async dequeueAll(): Promise<BatchItem[]> {
    const previousOp = this.operationQueue;
    let items: BatchItem[] = [];

    this.operationQueue = previousOp.then(async () => {
      items = [...this.queue];
      this.queue = [];
      this.clientOrder = 0;

      const result = await this.persist();
      if (!result.success) {
        logger.error('Failed to persist empty queue after dequeue');
      }
    });

    await this.operationQueue;
    return items;
  }

  clear(): Promise<void> {
    const previousOp = this.operationQueue;
    this.operationQueue = previousOp.then(async () => {
      const result = await removeItem(STORAGE_KEYS.OFFLINE_QUEUE);
      if (!result.success) {
        logger.error('Failed to clear queue from storage');
        return;
      }
      this.queue = [];
      this.clientOrder = 0;
    });
    return this.operationQueue;
  }

  getSize(): number {
    return this.queue.length;
  }

  private async persist() {
    return await setItem(STORAGE_KEYS.OFFLINE_QUEUE, this.queue);
  }
}

File: packages/swift/Sources/PhaseAnalytics/Core/Queue/OfflineQueue.swift
import Foundation

internal actor OfflineQueue {
    private var queue: [BatchItem] = []
    private var clientOrder = 0
    private let storage: StorageAdapter
    private var initialized = false

    private var operationQueue = Task<Void, Never> {}

    private static let maxQueueSize = 5000

    init(storage: StorageAdapter) {
        self.storage = storage
    }

    func initialize() async {
        let result = await storage.getItem(key: StorageKeys.offlineQueue) as Result<[BatchItem]?, Error>

        if case .success(let items) = result, let items = items, !items.isEmpty {
            queue = items
            let orders = items.map { $0.clientOrder }
            clientOrder = (orders.max() ?? 0) + 1
            logger.info("Loaded \(items.count) items from offline queue")
        } else if case .failure(let error) = result {
            logger.error("Failed to load offline queue. Storage unavailable.", error)
        }

        initialized = true
    }

    func enqueue(_ item: BatchItem) async {
        guard initialized else {
            logger.error("OfflineQueue not initialized. Call initialize() first.")
            return
        }

        let previousOp = operationQueue
        operationQueue = Task {
            await previousOp.value

            if queue.count >= Self.maxQueueSize {
                dropOldestSession()
            }

            let isRequeue: Bool

            switch item {
            case .device(_, let order, _),
                .session(_, let order, _),
                .event(_, let order, _),
                .ping(_, let order, _):
                isRequeue = order > 0
            }

            let newItem: BatchItem
            if isRequeue {
                newItem = item
            } else {
                switch item {
                case .device(let payload, _, let retry):
                    newItem = .device(payload: payload, clientOrder: clientOrder, retryCount: retry)
                case .session(let payload, _, let retry):
                    newItem = .session(payload: payload, clientOrder: clientOrder, retryCount: retry)
                case .event(let payload, _, let retry):
                    newItem = .event(payload: payload, clientOrder: clientOrder, retryCount: retry)
                case .ping(let payload, _, let retry):
                    newItem = .ping(payload: payload, clientOrder: clientOrder, retryCount: retry)
                }
                clientOrder += 1
            }

            queue.append(newItem)

            let result = await persist()
            if case .failure = result {
                logger.error("Failed to persist queue. Data may be lost on app crash.")
            }
        }

        await operationQueue.value
    }

    func dequeueAll() async -> [BatchItem] {
        let previousOp = operationQueue
        var items: [BatchItem] = []

        operationQueue = Task {
            await previousOp.value

            items = queue
            queue = []
            clientOrder = 0

            let result = await persist()
            if case .failure = result {
                logger.error("Failed to persist empty queue. Storage unavailable.")
            }
        }

        await operationQueue.value
        return items
    }

    func clear() async {
        let previousOp = operationQueue

        operationQueue = Task {
            await previousOp.value

            let result = await storage.removeItem(key: StorageKeys.offlineQueue)
            if case .failure = result {
                logger.error("Failed to clear queue. Check storage permissions.")
                return
            }

            queue = []
            clientOrder = 0
        }

        await operationQueue.value
    }

    func getSize() -> Int {
        return queue.count
    }

    private func dropOldestSession() {
        let sessionItems = queue.filter { item in
            if case .session = item { return true }
            return false
        }

        guard !sessionItems.isEmpty else {
            if !queue.isEmpty {
                queue.removeFirst()
                logger.info("Queue full (\(Self.maxQueueSize)). No sessions to drop, dropping oldest item.")
            }
            return
        }

        guard
            let oldestSession = sessionItems.min(by: {
                $0.clientOrder < $1.clientOrder
            })
        else { return }

        let oldestSessionId: String
        if case .session(let payload, _, _) = oldestSession {
            oldestSessionId = payload.sessionId
        } else {
            return
        }

        let initialLength = queue.count
        queue = queue.filter { item in
            if case .device = item { return true }

            if case .session(let payload, _, _) = item {
                return payload.sessionId != oldestSessionId
            }

            if case .event(let payload, _, _) = item {
                return payload.sessionId != oldestSessionId
            }

            if case .ping(let payload, _, _) = item {
                return payload.sessionId != oldestSessionId
            }

            return true
        }

        let droppedCount = initialLength - queue.count
        logger.info("Queue full (\(Self.maxQueueSize)). Dropped oldest session and \(droppedCount) related items.")
    }

    private func persist() async -> Result<Void, Error> {
        return await storage.setItem(key: StorageKeys.offlineQueue, value: queue)
    }
}
